src/server/pkg/storage/fileset/index_resolver.go

```diff
@@ -0,0 +1,52 @@
+package fileset
+
+import (
+	"context"
+
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset/index"
+)
+
+// NewIndexResolver ensures the indexes in the FileSource are correct
+// based on the content
+func NewIndexResolver(x FileSource) FileSource {
+	switch x := x.(type) {
+	case *mergeSource:
+		return &mergeResolver{
+			getReader: x.getReader,
+			s:         x.s,
+		}
+	default:
+		panic("cannot resolve indexes")
+	}
+}
+
+type mergeResolver struct {
+	s         *Storage
+	getReader func() (*MergeReader, error)
+}
+
+func (mr *mergeResolver) Iterate(ctx context.Context, cb func(File) error, stopBefore ...string) error {
+	mr1, err := mr.getReader()
+	if err != nil {
+		return err
+	}
+	mr2, err := mr.getReader()
+	if err != nil {
+		return err
+	}
+	w := mr.s.newWriter(ctx, "", WithNoUpload(func(idx *index.Index) error {
+		fmr, err := mr2.Next()
+		if err != nil {
+			return err
+		}
+		if fmr.Index().Path != idx.Path {
+			panic("different merge readers from getReader")
```

Comment:
I'm not sure how I feel about a panic here. I don't think we want to kill the pachd instance because we got into this state when doing a merge.

Reply:
I'm usually okay with panicking on logic errors.  But I think there are probably cases here where the semantic path could change in between calls, or other out of process things happen.